### PR TITLE
Add type {:map, type}

### LINKED ISSUE #1291: Add type {:map, type}
I created this issue per discussion on https://github.com/hexpm/hex_web/issues/205. An example of how this could be used is the following (inspired by HexWeb use case, but simplified):

```elixir
defmodule Package do
  use Ecto.Schema

  schema "packages" do
    field :name, :string
    field :links, {:map, :string}
  end
end

%Package{name: "Ecto", links: %{"GitHub" => "https://github.com/elixir-lang/ecto"}}
```

### PR DESCRIPTION
**Pull request by wojtekmach**
Closes #1291

### CODE REVIEW

File: lib/ecto/type.ex [hunk 1]
```diff
@@ -101,10 +101,10 @@ defmodule Ecto.Type do
   @typep base      :: :integer | :float | :boolean | :string | :map |
                       :binary | :decimal | :id | :binary_id |
                       :datetime | :date | :time | :any
-  @typep composite :: {:array, base} | {:embed, Ecto.Embedded.t} | {:in, base}
+  @typep composite :: {:array, base} | {:map, base} | {:embed, Ecto.Embedded.t} | {:in, base}
```
Comment (wojtekmach) [inline lib/ecto/type.ex]: we can have arbitrary complex types e.g.: `{:map, {:array, {:map, :map}}}`, I'm not sure how to represent them in typespec though

File: lib/ecto/type.ex [hunk 2]
```diff
@@ -634,6 +646,7 @@ defmodule Ecto.Type do
   defp of_base_type?(:binary, term),     do: is_binary(term)
   defp of_base_type?(:string, term),     do: is_binary(term)
   defp of_base_type?(:map, term),        do: is_map(term) and not Map.has_key?(term, :__struct__)
+  defp of_base_type?({:map, _}, _),      do: false # Always handled explicitly.
```
Comment (wojtekmach) [inline lib/ecto/type.ex]: I wasn't sure what to put here, so I just copied what we had for `{:array, type}`

File: lib/ecto/schema.ex [hunk 3]
```diff
@@ -133,8 +133,9 @@ defmodule Ecto.Schema do
   `{:array, inner_type}`  | `list`                  | `[value, value, value, ...]`
   `:decimal`              | [`Decimal`](https://github.com/ericmj/decimal) |
   `:map`                  | `map` |
+  `{:map, inner_type}`    | `map` |
```
Comment (wojtekmach) [inline lib/ecto/schema.ex]: to make the API smaller, I'd consider eventually deprecating `:map` with perhaps `{:map, :any}` as a replacement.

**Comment by wojtekmach** [comment]
I added an initial draft of the PR and some comments, I'm sure there's more work to do here. Please let me know if this is going into the right direction. Thanks!

I'll look at the build failure now (I haven't run it with sojourn_broker locally)

File: lib/ecto/type.ex (hunk 1 above)
Comment (josevalim) [inline lib/ecto/type.ex]: Can you use `{:map, t}`? Maybe that's what we should use in `{:array, _}` and `{:in, _}` as well.

File: lib/ecto/type.ex (hunk 2 above)
Comment (josevalim) [inline lib/ecto/type.ex]: That's correct.

**Comment by josevalim** [comment]
This looks great. There is just one comment regarding the typespecs and we need to fix the build.

I am not sure about removing `:map` in favor of `{:map , :any}`. Specially because `:any` is only allowed in virtual fields today, I am not sure if we want to promote it elsewhere.

**Comment by wojtekmach** [comment]
@josevalim I applied code review feedback (and some other random fixes) in df3a585. I also tried integrating it to hex_web and it passed the test suite. Let me know if anything else is missing!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**